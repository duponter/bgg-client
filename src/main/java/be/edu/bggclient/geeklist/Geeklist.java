package be.edu.bggclient.geeklist;

import java.time.LocalDateTime;
import java.util.List;

public class Geeklist {
    private final LocalDateTime postDate;
    private final LocalDateTime editDate;
    private final int thumbs;
    private final int itemCount;
    private final String username;
    private final String title;
    private final String description;
    private final List<GeeklistItem> items;

    private Geeklist(Builder builder) {
        this.postDate = builder.postDate;
        this.editDate = builder.editDate;
        this.thumbs = builder.thumbs;
        this.itemCount = builder.itemCount;
        this.username = builder.username;
        this.title = builder.title;
        this.description = builder.description;
        this.items = List.copyOf(builder.items);
    }

    public LocalDateTime getPostDate() {
        return postDate;
    }

    public LocalDateTime getEditDate() {
        return editDate;
    }

    public int getThumbs() {
        return thumbs;
    }

    public int getItemCount() {
        return itemCount;
    }

    public String getUsername() {
        return username;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public List<GeeklistItem> getItems() {
        return items;
    }

    public static final class Builder {
        private LocalDateTime postDate;
        private LocalDateTime editDate;
        private int thumbs;
        private int itemCount;
        private String username;
        private String title;
        private String description;
        private List<GeeklistItem> items;

        public Builder withPostDate(LocalDateTime postDate) {
            this.postDate = postDate;
            return this;
        }

        public Builder withEditDate(LocalDateTime editDate) {
            this.editDate = editDate;
            return this;
        }

        public Builder withThumbs(int thumbs) {
            this.thumbs = thumbs;
            return this;
        }

        public Builder withItemCount(int itemCount) {
            this.itemCount = itemCount;
            return this;
        }

        public Builder withUsername(String username) {
            this.username = username;
            return this;
        }

        public Builder withTitle(String title) {
            this.title = title;
            return this;
        }

        public Builder withDescription(String description) {
            this.description = description;
            return this;
        }

        public Builder withItems(List<GeeklistItem> items) {
            this.items = items;
            return this;
        }

        public Geeklist build() {
            return new Geeklist(this);
        }
    }
}
