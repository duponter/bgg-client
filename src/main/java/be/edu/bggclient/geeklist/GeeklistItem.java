package be.edu.bggclient.geeklist;

import java.time.LocalDateTime;

public class GeeklistItem {
    private final String id;
    private final String objectType;
    private final String subType;
    private final String objectId;
    private final String objectName;
    private final String username;
    private final LocalDateTime postDate;
    private final LocalDateTime editDate;
    private final int thumbs;
    private final String imageId;
    private final String comments;

    public String getId() {
        return id;
    }

    public String getObjectType() {
        return objectType;
    }

    public String getSubType() {
        return subType;
    }

    public String getObjectId() {
        return objectId;
    }

    public String getObjectName() {
        return objectName;
    }

    public String getUsername() {
        return username;
    }

    public LocalDateTime getPostDate() {
        return postDate;
    }

    public LocalDateTime getEditDate() {
        return editDate;
    }

    public int getThumbs() {
        return thumbs;
    }

    public String getImageId() {
        return imageId;
    }

    public String getComments() {
        return comments;
    }

    public GeeklistItem(Builder builder) {
        this.id = builder.id;
        this.objectType = builder.objectType;
        this.subType = builder.subType;
        this.objectId = builder.objectId;
        this.objectName = builder.objectName;
        this.username = builder.username;
        this.postDate = builder.postDate;
        this.editDate = builder.editDate;
        this.thumbs = builder.thumbs;
        this.imageId = builder.imageId;
        this.comments = builder.comments;
    }

    public static final class Builder {
        private String id;
        private String objectType;
        private String subType;
        private String objectId;
        private String objectName;
        private String username;
        private LocalDateTime postDate;
        private LocalDateTime editDate;
        private int thumbs;
        private String imageId;
        private String comments;

        public Builder withId(String id) {
            this.id = id;
            return this;
        }

        public Builder withObjectType(String objectType) {
            this.objectType = objectType;
            return this;
        }

        public Builder withSubType(String subType) {
            this.subType = subType;
            return this;
        }

        public Builder withObjectId(String objectId) {
            this.objectId = objectId;
            return this;
        }

        public Builder withObjectName(String objectName) {
            this.objectName = objectName;
            return this;
        }

        public Builder withUsername(String username) {
            this.username = username;
            return this;
        }

        public Builder withPostDate(LocalDateTime postDate) {
            this.postDate = postDate;
            return this;
        }

        public Builder withEditDate(LocalDateTime editDate) {
            this.editDate = editDate;
            return this;
        }

        public Builder withThumbs(int thumbs) {
            this.thumbs = thumbs;
            return this;
        }

        public Builder withImageId(String imageId) {
            this.imageId = imageId;
            return this;
        }

        public Builder withComments(String comments) {
            this.comments = comments;
            return this;
        }

        public GeeklistItem build() {
            return new GeeklistItem(this);
        }
    }
}
