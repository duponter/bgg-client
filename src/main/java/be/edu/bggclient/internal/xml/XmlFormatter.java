package be.edu.bggclient.internal.xml;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

public final class XmlFormatter {
    private static final DateTimeFormatter BGG_DATE_TIME = DateTimeFormatter.ofPattern("E, dd MMM yyyy HH:mm:ss +0000", Locale.US);

    private XmlFormatter() {
        throw new UnsupportedOperationException("This static class cannot be instantiated.");
    }

    public static LocalDate parseDate(String input) {
        return LocalDate.parse(input);
    }

    public static LocalDateTime parseDateTime(String input) {
        return LocalDateTime.parse(input, BGG_DATE_TIME);
    }

    public static String format(String input) {
        return input;
    }

    public static String format(int input) {
        return Integer.toString(input);
    }

    public static String format(boolean input) {
        return input ? "1" : "0";
    }

    public static String format(LocalDate input) {
        return input.toString();
    }

    public static String format(LocalDateTime input) {
        return input.format(BGG_DATE_TIME);
    }
}
