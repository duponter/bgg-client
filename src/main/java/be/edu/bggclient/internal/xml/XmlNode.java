package be.edu.bggclient.internal.xml;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Objects;
import java.util.stream.IntStream;
import java.util.stream.Stream;
import javax.xml.namespace.QName;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

public abstract class XmlNode {
    private final XPath xpath = XPathFactory.newInstance().newXPath();
    private final Node node;

    public static Stream<Node> nodes(Node root, String expression) {
        XmlNode rootNode = new XmlNode(root) {
        };
        return rootNode.nodes(expression).map(XmlNode::detach);
    }

    private static Node detach(Node node) {
        //https://stackoverflow.com/questions/3782618/xpath-evaluate-performance-slows-down-absurdly-over-multiple-calls
        if (node.getParentNode() != null) {
            node.getParentNode().removeChild(node);
        }
        return node;
    }

    protected XmlNode(Node node) {
        this.node = Objects.requireNonNull(node);
    }

    protected String stringValueAttribute(String attribute) {
        return this.string(attribute + "/@value");
    }

    protected String string(String expression) {
        return (String) this.data(expression, XPathConstants.STRING);
    }

    protected LocalDateTime dateTime(String expression) {
        return LocalDateTime.parse(this.string(expression), DateTimeFormatter.RFC_1123_DATE_TIME);
    }

    protected Number numericValueAttribute(String attribute) {
        return this.number(attribute + "/@value");
    }

    protected Number number(String expression) {
        return (Number) this.data(expression, XPathConstants.NUMBER);
    }

    protected boolean toBoolean(String expression) {
        return number(expression).intValue() != 0;
    }

    protected Stream<Node> nodes(String expression) {
        NodeList nodeList = (NodeList) this.data(expression, XPathConstants.NODESET);
        return IntStream.range(0, nodeList.getLength()).mapToObj(nodeList::item);
    }

    private Object data(String expression, QName returnType) {
        try {
            return xpath.evaluate(expression, node, returnType);
        } catch (XPathExpressionException e) {
            throw new IllegalStateException("Unable to evaluate xpath expression: " + expression, e);
        }
    }
}
