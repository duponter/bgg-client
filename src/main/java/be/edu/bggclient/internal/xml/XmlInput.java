package be.edu.bggclient.internal.xml;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

public final class XmlInput {
    private static final DocumentBuilderFactory FACTORY = DocumentBuilderFactory.newInstance();

    static {
        FACTORY.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        FACTORY.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
    }

    private XmlInput() {
        throw new UnsupportedOperationException("This static class cannot be instantiated.");
    }

    public static DocumentBuilder newDocumentBuilder() {
        try {
            return FACTORY.newDocumentBuilder();
        } catch (ParserConfigurationException pce) {
            throw new IllegalStateException("Unable to create DocumentBuilder from factory", pce);
        }
    }
}
