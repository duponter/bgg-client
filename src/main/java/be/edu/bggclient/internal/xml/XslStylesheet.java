package be.edu.bggclient.internal.xml;

import java.io.IOException;
import java.io.StringWriter;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;
import javax.xml.XMLConstants;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;

import org.w3c.dom.Node;

public class XslStylesheet {
    private static final String XML_DECLARATION = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>";
    private static final String XSL_DIRECTIVE = "<?xml-stylesheet version=\"2.0\" type=\"text/xsl\" href=\"%s\"?>";

    private final String location;

    public XslStylesheet(String location) {
        this.location = location;
    }

    public String includeInXml(String xml) {
        return String.join(
                System.lineSeparator(),
                XML_DECLARATION,
                String.format(XSL_DIRECTIVE, location),
                xml.replace(XML_DECLARATION, "")
        );
    }

    public String read() {
        String filename = this.asFilename();
        try {
            return Files.readString(Path.of(Objects.requireNonNull(XslStylesheet.class.getResource(filename)).toURI()));
        } catch (IOException | URISyntaxException e) {
            throw new IllegalStateException("Unable to read Xsl Stylesheet " + filename, e);
        }
    }

    public String apply(Node node) {
        TransformerFactory factory = TransformerFactory.newInstance();
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");

        try {
            Transformer transformer = factory.newTransformer(new StreamSource(XslStylesheet.class.getResourceAsStream(this.asFilename())));
            StringWriter stringWriter = new StringWriter();
            transformer.transform(new DOMSource(node), new StreamResult(stringWriter));
            return stringWriter.toString();
        } catch (TransformerException e) {
            e.printStackTrace();
            throw new IllegalStateException(e);
        }

    }

    private String asFilename() {
        int pos = this.location.lastIndexOf("/");
        return (pos == -1 ? this.location : this.location.substring(0, pos)) + ".xsl";
    }

}
