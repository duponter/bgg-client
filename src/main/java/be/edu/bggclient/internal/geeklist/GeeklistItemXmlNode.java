package be.edu.bggclient.internal.geeklist;

import be.edu.bggclient.geeklist.GeeklistItem;
import be.edu.bggclient.internal.Builder;
import be.edu.bggclient.internal.xml.XmlNode;
import org.w3c.dom.Node;

public class GeeklistItemXmlNode extends XmlNode implements Builder<GeeklistItem> {
    public GeeklistItemXmlNode(Node node) {
        super(node);
    }

    @Override
    public GeeklistItem build() {
        return new GeeklistItem.Builder()
                .withId(this.string("@id"))
                .withObjectType(this.string("@objecttype"))
                .withSubType(this.string("@subtype"))
                .withObjectId(this.string("@objectid"))
                .withObjectName(this.string("@objectname"))
                .withUsername(this.string("@username"))
                .withPostDate(this.dateTime("@postdate"))
                .withEditDate(this.dateTime("@editdate"))
                .withThumbs(this.number("@thumbs").intValue())
                .withImageId(this.string("@imageid"))
                .withComments(this.string("body"))
                .build();
    }
}
