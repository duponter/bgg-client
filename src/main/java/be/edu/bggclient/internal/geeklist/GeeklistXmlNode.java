package be.edu.bggclient.internal.geeklist;

import java.util.stream.Collectors;

import be.edu.bggclient.geeklist.Geeklist;
import be.edu.bggclient.internal.Builder;
import be.edu.bggclient.internal.xml.XmlFormatter;
import be.edu.bggclient.internal.xml.XmlNode;
import org.w3c.dom.Node;

public class GeeklistXmlNode extends XmlNode implements Builder<Geeklist> {
    public GeeklistXmlNode(Node node) {
        super(node);
    }

    @Override
    public Geeklist build() {
        return new Geeklist.Builder()
                .withId(this.string("@id"))
                .withPostDate(XmlFormatter.parseDateTime(this.string("postdate")))
                .withEditDate(XmlFormatter.parseDateTime(this.string("editdate")))
                .withThumbs(this.number("thumbs").intValue())
                .withItemCount(this.number("numitems").intValue())
                .withUsername(this.string("username"))
                .withTitle(this.string("title"))
                .withDescription(this.string("description"))
                .withItems(
                        this.nodes("item")
                                .map(GeeklistItemXmlNode::new)
                                .map(Builder::build)
                                .collect(Collectors.toList())
                )
                .build();
    }
}
