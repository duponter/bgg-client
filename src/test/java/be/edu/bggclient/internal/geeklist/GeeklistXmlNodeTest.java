package be.edu.bggclient.internal.geeklist;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

import be.edu.bggclient.geeklist.Geeklist;
import be.edu.bggclient.geeklist.GeeklistItem;
import be.edu.bggclient.internal.xml.XmlInput;
import be.edu.bggclient.internal.xml.XmlNode;
import be.edu.bggclient.internal.xml.XslStylesheet;
import org.approvaltests.JsonApprovals;
import org.dom4j.DocumentHelper;
import org.dom4j.Element;
import org.dom4j.io.OutputFormat;
import org.dom4j.io.XMLWriter;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class GeeklistXmlNodeTest {
    @Test
    void mapsXmlToGeeklistPojo() throws IOException {
        try (InputStream xml = GeeklistXmlNodeTest.class.getResourceAsStream("geeklist.xml")) {
            assertThat(xml).isNotNull();
            // TODO_EDU support multiple geeklists?
            Geeklist geeklist = XmlNode.nodes(new XmlInput().read(xml), "//geeklist")
                    .map(GeeklistXmlNode::new)
                    .map(GeeklistXmlNode::build)
                    .findFirst()
                    .orElseThrow();
            JsonApprovals.verifyAsJson(geeklist);

            OutputFormat format = OutputFormat.createPrettyPrint();
            XMLWriter writer = new XMLWriter(System.out, format);
            writer.write(toXml(geeklist));
        }
    }

    private Element toXml(Geeklist geeklist) {
        Element geeklistNode = DocumentHelper.createElement("geeklist");
        geeklistNode.addAttribute("id", geeklist.getId());
        geeklistNode.addElement("postdate").addText(geeklist.getPostDate().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.RFC_1123_DATE_TIME));
        geeklistNode.addElement("editdate").addText(geeklist.getEditDate().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.RFC_1123_DATE_TIME));
        geeklistNode.addElement("thumbs").addText(Integer.toString(geeklist.getThumbs()));
        geeklistNode.addElement("numitems").addText(Integer.toString(geeklist.getItemCount()));
        geeklistNode.addElement("username").addText(geeklist.getUsername());
        geeklistNode.addElement("title").addText(geeklist.getTitle());
        geeklistNode.addElement("description").addText(geeklist.getDescription());
        geeklist.getItems().forEach(geeklistItem -> geeklistNode.add(toXml(geeklistItem)));
        return geeklistNode;
    }

    private Element toXml(GeeklistItem item) {
        Element itemNode = DocumentHelper.createElement("item");
        itemNode.addAttribute("id", item.getId());
        itemNode.addAttribute("objecttype", item.getObjectType());
        itemNode.addAttribute("subtype", item.getSubType());
        itemNode.addAttribute("objectid", item.getObjectId());
        itemNode.addAttribute("objectname", item.getObjectName());
        itemNode.addAttribute("username", item.getUsername());
        itemNode.addAttribute("postdate", item.getPostDate().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.RFC_1123_DATE_TIME));
        itemNode.addAttribute("editdate", item.getEditDate().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.RFC_1123_DATE_TIME));
        itemNode.addAttribute("thumbs", Integer.toString(item.getThumbs()));
        itemNode.addAttribute("imageid", item.getImageId());
        itemNode.addElement("body").addText(item.getComments());
        return itemNode;
    }

    @Test
    void generatesJsonFromXml() throws IOException {
        try (InputStream xml = GeeklistXmlNodeTest.class.getResourceAsStream("geeklist.xml");
             InputStream expected = GeeklistXmlNodeTest.class.getResourceAsStream("GeeklistXmlNodeTest.mapsXmlToGeeklistPojo.approved.json")) {
            XslStylesheet xslStylesheet = new XslStylesheet("/be/edu/bggclient/internal/geeklist/geeklist-to-json/xsl");
            String toJson = xslStylesheet.apply(XmlNode.nodes(new XmlInput().read(xml), "//geeklist").findFirst().orElseThrow());
            System.out.println(toJson);
            assertThat(expected).asString(Charset.defaultCharset()).isEqualToIgnoringWhitespace(toJson);
        }
    }
}
