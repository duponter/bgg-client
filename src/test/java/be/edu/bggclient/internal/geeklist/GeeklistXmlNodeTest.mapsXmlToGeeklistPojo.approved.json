{
  "id": "274761",
  "postDate": "2020-07-15T07:17:21",
  "editDate": "2020-11-11T09:05:34",
  "thumbs": 0,
  "itemCount": 10,
  "username": "duponter",
  "title": "Borrowed Board Games",
  "description": "Board Games temporarily in my possession.",
  "items": [
    {
      "id": "7765962",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "160495",
      "objectName": "ZhanGuo",
      "username": "duponter",
      "postDate": "2020-07-15T07:16:09",
      "editDate": "2020-07-15T07:16:09",
      "thumbs": 0,
      "imageId": "2242246",
      "comments": "Forum Federatie"
    },
    {
      "id": "9153294",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "270673",
      "objectName": "Silver \u0026 Gold",
      "username": "duponter",
      "postDate": "2022-08-26T19:39:49",
      "editDate": "2022-08-26T19:39:49",
      "thumbs": 0,
      "imageId": "6974604",
      "comments": "Forum Mortsel"
    },
    {
      "id": "9153298",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "300531",
      "objectName": "Paleo",
      "username": "duponter",
      "postDate": "2022-08-26T19:40:06",
      "editDate": "2022-08-26T19:40:06",
      "thumbs": 0,
      "imageId": "6039256",
      "comments": "Forum Mortsel"
    },
    {
      "id": "9153300",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "43570",
      "objectName": "Friday",
      "username": "duponter",
      "postDate": "2022-08-26T19:40:23",
      "editDate": "2022-08-26T19:40:23",
      "thumbs": 0,
      "imageId": "1513328",
      "comments": "Forum Mortsel"
    },
    {
      "id": "9265058",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "339906",
      "objectName": "The Hunger",
      "username": "duponter",
      "postDate": "2022-10-09T20:47:12",
      "editDate": "2022-10-09T20:47:12",
      "thumbs": 0,
      "imageId": "6218039",
      "comments": "Forum Mortsel"
    },
    {
      "id": "9265061",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "346965",
      "objectName": "Azul: Queen\u0027s Garden",
      "username": "duponter",
      "postDate": "2022-10-09T20:47:34",
      "editDate": "2022-10-09T20:47:34",
      "thumbs": 0,
      "imageId": "6389125",
      "comments": "Forum Mortsel"
    },
    {
      "id": "9293548",
      "objectType": "thing",
      "subType": "boardgame",
      "objectId": "284378",
      "objectName": "Kanban EV",
      "username": "duponter",
      "postDate": "2022-10-24T10:56:12",
      "editDate": "2022-10-24T10:56:12",
      "thumbs": 0,
      "imageId": "4924232",
      "comments": "Bart"
    },
    {
      "id": "9302273",
      "objectType": "company",
      "subType": "boardgamepublisher",
      "objectId": "597",
      "objectName": "Eagle-Gryphon Games",
      "username": "duponter",
      "postDate": "2022-10-28T16:13:41",
      "editDate": "2022-10-28T16:14",
      "thumbs": 0,
      "imageId": "4031835",
      "comments": "Publisher as an example"
    },
    {
      "id": "9302277",
      "objectType": "person",
      "subType": "boardgamedesigner",
      "objectId": "11767",
      "objectName": "Alexander Pfister",
      "username": "duponter",
      "postDate": "2022-10-28T16:14:25",
      "editDate": "2022-10-28T16:14:25",
      "thumbs": 0,
      "imageId": "1079358",
      "comments": "Designer as an example"
    },
    {
      "id": "9302280",
      "objectType": "person",
      "subType": "boardgameartist",
      "objectId": "70533",
      "objectName": "Mihajlo Dimitrievski",
      "username": "duponter",
      "postDate": "2022-10-28T16:14:45",
      "editDate": "2022-10-28T16:14:45",
      "thumbs": 0,
      "imageId": "1774562",
      "comments": "Artist as an example"
    }
  ]
}
